src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1013 @@
+
+  Inputs:
+   index_expr - The index expression to  to resolve
+   out_expressions - The expressions this expression could be
+   out_is_const - Is the squashed expression constant
+
+ Outputs: Returns true if it was able to squash the index expression
+          If this is the case, out_expressions will contain
+          the possible values this index_of could return
+          The out_is_const will return whether either the array itself
+          is const, or the values of the array are const.
+
+ Purpose: To squash an index access by first finding the array it is accessing
+          Then if the index can be resovled, return the squashed value. If
+          the index can't be determined then squash each value in the array
+          and return them all.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::try_resolve_index_of(
+  const index_exprt &index_expr,
+  expressionst &out_expressions,
+  bool &out_is_const)
+{
+  // Get the array(s) it belongs to
+  expressionst potential_array_exprs;
+  bool array_const=false;
+  bool resolved_array=
+    try_resolve_expression(
+      index_expr.array(),
+      potential_array_exprs,
+      array_const);
+
+  if(resolved_array)
+  {
+    bool all_possible_const=true;
+    for(const exprt &potential_array_expr : potential_array_exprs)
+    {
+      all_possible_const=
+        all_possible_const &&
+        is_const_type(potential_array_expr.type().subtype());
+
+      if(potential_array_expr.id()==ID_array)
+      {
+        // Get the index if we can
+        mp_integer value;
+        if(try_resolve_index_value(index_expr.index(), value))
+        {
+          expressionst array_out_functions;
+          const exprt &func_expr=
+            potential_array_expr.operands()[integer2size_t(value)];
+          bool value_const=false;
+          bool resolved_value=
+            try_resolve_expression(func_expr, array_out_functions, value_const);
+
+          if(resolved_value)
+          {
+            out_expressions.insert(
+              out_expressions.end(),
+              array_out_functions.begin(),
+              array_out_functions.end());
+          }
+          else
+          {
+            LOG("Failed to resolve array value", func_expr);
+            return false;
+          }
+        }
+        else
+        {
+          // We don't know what index it is,
+          // but we know the value is from the array
+          for(const exprt &array_entry : potential_array_expr.operands())
+          {
+            expressionst array_contents;
+            bool is_entry_const;
+            bool resolved_value=
+              try_resolve_expression(
+                array_entry, array_contents, is_entry_const);
+
+            if(resolved_value)
+            {
+              for(const exprt &resolved_array_entry : array_contents)
+              {
+                if(resolved_array_entry .is_zero())
```

Comment:
Removing braces introduce bugs like this:https://github.com/diffblue/cbmc-toyota/pull/63#pullrequestreview-27401804 so I'd really rather not do this...